docs/source/cc_launcher.md

```diff
@@ -0,0 +1,301 @@
+chaincode was part of the peer implementation and could not be easily
+customized. All chaincode instantiated on the peer would be "built" using
+language specific logic hard-coded in the peer. This build process would
+generate a Docker container image that would be launched to execute chaincode
+that connected as a client to the peer.
+
+This approach limited chaincode implementations to a handful of languages,
+required Docker to be part of the deployment environment, and prevented
+running chaincode as a long running server process.
+
+External Builders and Launchers address these limitations by enabling
+operators to extend the peer with programs that can build, launch, and
+discover chaincode.
+
+## External Builder Model
+
+Hyperledger Fabric External Builders and Launchers are loosely based on Heroku
+[Buildpacks][buildpacks]. A buildpack implementation is simply a collection of
+programs or scripts that transform application artifacts into something that
+can run. The buildpack model has been adapted for chaincode packages and
+extended to support chaincode execution and discovery.
+
+### External Builder and Launcher API
+
+An external builder and launcher consists of four programs or scripts:
+
+- `bin/detect`: Determine whether or not this buildpack should be used to
+  build the chaincode package and launch it.
+- `bin/build`: Transform the chaincode package into executable chaincode.
+- `bin/release`: Provide metadata to the peer about the chaincode. (optional)
+- `bin/run`: Run the chaincode.  (optional)
+
+#### `bin/detect`
+
+The `bin/detect` script is responsible for determining whether or not a buildpack
+should be used to build a chaincode package and launch it. The peer invokes
+`detect` with two arguments:
+
+```sh
+bin/detect CHAINCODE_SOURCE_DIR CHAINCODE_METADATA_DIR
+```
+
+When `detect` is invoked, `CHAINCODE_SOURCE_DIR` contains the chaincode source
+and `CHAINCODE_METADATA_DIR` contains the `metadata.json` file from the
+chaincode package installed to the peer. If the buildpack should be applied to
+the chaincode source package, `detect` must return an exit code of `0`; any
+other exit code will indicate that the buildpack should not be applied.
+
+The following is an example of a simple `detect` script for go chaincode:
+```sh
+#!/bin/bash
+
+CHAINCODE_METADATA_DIR="$2"
+
+# use jq to extract the chaincode type from metadata.json and exit with
+# success if the chaincode type is golang
+if [ "$(jq -r .type "$CHAINCODE_METADATA_DIR/metadata.json" | tr '[:upper:]' '[:lower:]')" = "golang" ]; then
+    exit 0
+fi
+
+exit 1
+```
+
+#### `bin/build`
+
+The `bin/build` script is responsible for building, compiling, or transforming
+the contents of a chaincode package into artifacts that can be used by
+`release` and `run`. The peer invokes `build` with three arguments:
+
+```sh
+bin/build CHAINCODE_SOURCE_DIR CHAINCODE_METADATA_DIR BUILD_OUTPUT_DIR
+```
```

Comment:
Perhaps note that the source dir and metadata dir should be treated as read only inputs, and the output dir as writable.

Reply:
Done.